packages/datagrid/src/imagerenderer.ts

```diff
@@ -0,0 +1,252 @@
+   */
+  paintPlaceholder(gc: GraphicsContext, config: CellRenderer.CellConfig): void {
+    this.drawBackground(gc, config);
+    this.drawPlaceholder(gc, config);
+  }
+
+  /**
+   * Paint the content for a cell.
+   *
+   * @param gc - The graphics context to use for drawing.
+   *
+   * @param config - The configuration data for the cell.
+   */
+  paint(gc: GraphicsContext, config: CellRenderer.CellConfig): void {
+    this.drawBackground(gc, config);
+    this.drawImage(gc, config);
+  }
+
+  /**
+   * Draw the background for the cell.
+   *
+   * @param gc - The graphics context to use for drawing.
+   *
+   * @param config - The configuration data for the cell.
+   */
+  drawBackground(gc: GraphicsContext, config: CellRenderer.CellConfig): void {
+    // Resolve the background color for the cell.
+    const color = CellRenderer.resolveOption(this.backgroundColor, config);
+
+    // Bail if there is no background color to draw.
+    if (!color) {
+      return;
+    }
+
+    // Fill the cell with the background color.
+    gc.fillStyle = color;
+    gc.fillRect(config.x, config.y, config.width, config.height);
+  }
+
+  /**
+   * Draw the placeholder for the cell.
+   *
+   * @param gc - The graphics context to use for drawing.
+   *
+   * @param config - The configuration data for the cell.
+   */
+  drawPlaceholder(gc: GraphicsContext, config: CellRenderer.CellConfig): void {
+    const placeholder = CellRenderer.resolveOption(this.placeholder, config);
+    const color = CellRenderer.resolveOption(this.textColor, config);
+
+    const textX = config.x + config.width / 2;
+    const textY = config.y + config.height / 2;
+
+    // Draw the placeholder.
+    gc.fillStyle = color;
+    gc.fillText(placeholder, textX, textY);
+  }
+
+  /**
+   * Draw the image for the cell.
+   *
+   * @param gc - The graphics context to use for drawing.
+   *
+   * @param config - The configuration data for the cell.
+   */
+  drawImage(gc: GraphicsContext, config: CellRenderer.CellConfig): void {
+    // Bail early if there is nothing to draw
+    if (!config.value) {
+      return;
+    }
+
+    const img = ImageRenderer.dataCache.get(config.value);
+
+    // If it's not loaded yet, show the placeholder
+    if (!img) {
+      return this.drawPlaceholder(gc, config);
+    }
+
+    const fitWidth = (img.width / img.height) * config.height;
+    const fitHeight = (img.height / img.width) * config.width;
+    switch (this.sizingMode) {
+      case 'fit-height':
+        gc.drawImage(img, config.x, config.y, fitWidth, config.height);
+        break;
+      case 'fit-width':
+        gc.drawImage(img, config.x, config.y, config.width, fitHeight);
+        break;
+      case 'fill':
+        gc.drawImage(img, config.x, config.y, config.width, config.height);
+        break;
+      case 'original':
+        gc.drawImage(img, config.x, config.y);
+        break;
+    }
+  }
+
+  static dataCache = new Map<string, HTMLImageElement | undefined>();
```

Comment:
Do elements in this cache need to be disposed some way to avoid memory leaks?

Reply:
As it's a static property, it will live as long as the page is running, is that right? So I guess there will be no memory leak happening?

Reply:
I thought we may want to clean the cache periodically, could be done in a separate PR if needed.

Reply:
In context of outputs in a notebook, if user opens and then closes 100 notebooks with different in images in data grids, will the memory stay constant or increase with each notebook? If this was tied to a renderer instance, I imagine we would not need to worry as the reference would not linger, but then we would need to fetch the resource for each output rendering which is not ideal either.

Reply:
Culling the cache periodically might be a solution indeed, but we would want to do this selectively (not touch any currently rendered data grids). So we may want to change the cache implementation in the future to do some reference counting or time-based expiration. To make it easier on future ourselves, what if that was a non-exported global variable so that it is not a part of public API? Any downside to this idea?

Reply:
Memory will increase if those images are always different.

> To make it easier on future ourselves, what if that was a non-exported global variable so that it is not a part of public API

Right!! We should definitely make that cache property private.

I'm fine with the idea to periodically clean the cache. The browser will have cached the image somewhere anyway, so next time we reload the same image it should not take much time to download again.

Reply:
Yes, absolutely!